contracts/src/v0.1/L2endpoint.sol

```diff
@@ -53,4 +72,109 @@ contract L2Endpoint is Ownable {
     }
+
+    function computeRequestId(
+        bytes32 keyHash,
+        address sender,
+        uint64 accId,
+        uint64 nonce
+    ) private pure returns (uint256, uint256) {
+        uint256 preSeed = uint256(keccak256(abi.encode(keyHash, sender, accId, nonce)));
+        uint256 requestId = uint256(keccak256(abi.encode(keyHash, preSeed)));
+        return (requestId, preSeed);
+    }
+
+    function requestRandomWords(
+        bytes32 keyHash,
+        uint64 accId,
+        uint32 callbackGasLimit,
+        uint32 numWords,
+        bool isDirectPayment
+    ) private returns (uint256) {
+        sNonce++;
+        (uint256 requestId, uint256 preSeed) = computeRequestId(keyHash, msg.sender, accId, sNonce);
+        sRequestDetail[requestId] = RequestInfo({
+            owner: msg.sender,
+            callbackGasLimit: callbackGasLimit
+        });
+        emit RandomWordsRequested(
+            keyHash,
+            requestId,
+            preSeed,
+            accId,
+            callbackGasLimit,
+            numWords,
+            msg.sender,
+            isDirectPayment
+        );
+
+        return requestId;
+    }
+
+    /**
+     * @dev calls target address with exactly gasAmount gas and data as calldata
+     * or reverts if at least gasAmount gas is not available.
+     */
+    function callWithExactGas(
+        uint256 gasAmount,
+        address target,
+        bytes memory data
+    ) internal returns (bool success) {
+        // solhint-disable-next-line no-inline-assembly
+        assembly {
+            let g := gas()
+            // Compute g -= GAS_FOR_CALL_EXACT_CHECK and check for underflow
+            // The gas actually passed to the callee is min(gasAmount, 63//64*gas available).
+            // We want to ensure that we revert if gasAmount >  63//64*gas available
+            // as we do not want to provide them with less, however that check itself costs
+            // gas.  GAS_FOR_CALL_EXACT_CHECK ensures we have at least enough gas to be able
+            // to revert if gasAmount >  63//64*gas available.
+            if lt(g, GAS_FOR_CALL_EXACT_CHECK) {
+                revert(0, 0)
+            }
+            g := sub(g, GAS_FOR_CALL_EXACT_CHECK)
+            // if g - g//64 <= gasAmount, revert
+            // (we subtract g//64 because of EIP-150)
+            if iszero(gt(sub(g, div(g, 64)), gasAmount)) {
+                revert(0, 0)
+            }
+            // solidity calls check that a contract actually exists at the destination, so we do the same
+            if iszero(extcodesize(target)) {
+                revert(0, 0)
+            }
+            // call and return whether we succeeded. ignore return data
+            // call(gas,addr,value,argsOffset,argsLength,retOffset,retLength)
+            success := call(gasAmount, target, 0, add(data, 0x20), mload(data), 0, 0)
+        }
+        return success;
+    }
+
+    function requestRandomWords(
```

Comment:
Since we do not need to distinguish between direct payment and prepayment, we do not need to split request to two functions. We can call directly the one above.